### ISSUE TITLE
Make use of VirtualTrees.StyleHooks optional

### ISSUE THREAD

**Issue by pyscripter** [JAM-Software/Virtual-TreeView#532]
a) Move 

``` Delphi
  //Note - not using class constructors as they are not supported on C++ Builder.
  TCustomStyleEngine.RegisterStyleHook(TVirtualStringTree, TVclStyleScrollBarsHook);
  TCustomStyleEngine.RegisterStyleHook(TVirtualDrawTree, TVclStyleScrollBarsHook);
```

from the initialization part of VirtualTrees to the initialization part of VirtualTrees.StyleHooks

b) Remove VirtualTrees.StyleHooks from the uses list of VirtualTrees.pas

Rational
Users have a choice of including the hooks, just by adding VirtualTrees.StyleHooks to the project file.

Notes
-  The hooks are supposed to style the scrollbars (I think), but I see no visual difference with/out the hooks (XE7).
-  In a styled application I saw exceptions raised in the code of the style hooks (invalid window handle), which are very hard to debug or reproduce.
-  Perfect styling of the scrollbars can be achieved by adding to the project the unit Vcl.Styles.UxTheme from https://code.google.com/p/vcl-styles-utils/.

**Comment by joachimmarder** [comment]
We tried to do the VCL styles very similar as they are done for VCL controls, e.g. TTreeView. This is why I don't think it is a good idea to makes the registration of the hooks optional. As far as I know they are only used when needed.

But I have to admit that I don't use VCL styles and so I am not an expert. If more people support this change, we can change this.

Using Vcl.Styles.UxTheme sounds like an interesting idea, licenses should be compatible.

**Comment by vincentparrett** [comment]
They stylehooks are working fine for me, in an application that does use vcl styles. If you use the windows theme then the stylehooks do nothing. As for the vcl style utils, I'm not using them, I did try them but had all sorts of problems. I don't think we need to take a dependency on another project. 

These days, all vcl controls should support styling, it's not just about stylehooks (that is only part of it). 

The way stylle hooks are registered is a pain, but that's how emabarcadero designed it to work.

**Comment by pyscripter** [comment]
I am not suggesting to create a dependency to other projects.   I just mentioned, that if you want proper styling of scrollbars in your projects you may want to use the units from the VCL styles project, which by the way work fine for me.

The issues I reported are:
-  The style hooks of VirtualTree don't style the scrollbars properly, at least not for me.  In fact I see no difference with/out the hooks with an active VCL style (Jet, Carbon etc.).
-  I experienced hard to debug crashes in the hooks code (invalid windows handle)
-  Even VCL allows you to selectively disable certain hooks.  VirtualTree should do the same, i.e make their use optional.   The easiest way to achieve these without introducing new options, flags etc. is by implementing the simple changes I originally suggested.

**Comment by joachimmarder** [comment]
> The style hooks of VirtualTree don't style the scrollbars properly, at least not for me. In fact I see no
> difference with/out the hooks with an active VCL style (Jet, Carbon etc.).

There are no known problems.
Can ypu please attach a screenshot of one of our sample projects which shows the issue?

> I experienced hard to debug crashes in the hooks code (invalid windows handle)

Without steps for reproduction or a sample project there is not much we can do. A report generated with MadExcept using one of the included sample projects could help too.

> Even VCL allows you to selectively disable certain hooks.

Can you please provide more details to what you refer exactly?

**Comment by pyscripter** [comment]
Virtual-Tree , Jet VCL Style, Hooks disabled (RegisterStyleHook lines commmented out.
![image](https://cloud.githubusercontent.com/assets/1311616/6896493/b08c0a34-d6f4-11e4-8a82-ebbe877ce6e1.png)

Virtual-Tree , Jet VCL Style, Hooks enabled.
![image](https://cloud.githubusercontent.com/assets/1311616/6896536/0a3f996a-d6f5-11e4-8eb9-154abf6b20f0.png)

Scrollbars of a standard TlistBox.
![image](https://cloud.githubusercontent.com/assets/1311616/6896583/89496e70-d6f5-11e4-8d63-a6fd1b96fcdd.png)

Virtual-Tree with  Vcl.Styles.UxTheme (unit from the VCL Styles project) added to the project.
![image](https://cloud.githubusercontent.com/assets/1311616/6896638/23c4e506-d6f6-11e4-8817-8d089921dac2.png)

**Comment by pyscripter** [comment]
An example of VCL disabling system hooks from VCL.Styles.

``` pascal
  TSystemHook = (shMenus); // Enumeration elements will be expanded in the future
  TSystemHooks = set of TSystemHook;
```

By specifying 

``` pascal
   TSystemHooks = [];
```

you can disable the menu hooks

**Comment by joachimmarder** [comment]
> Virtual-Tree , Jet VCL Style, Hooks enabled.

I cannot replicate this with our Advanced sample project and the included VCL styles. Scrollbars look as supposed here.

**Comment by joachimmarder** [comment]
> An example of VCL disabling system hooks from VCL.Styles.

Unit "Vcl.Styles" does not include such a code here. And this code looks entirely different from what you proposed: Moving the calls to RegisterStyleHook() to a place where they are bot called by default. VCL does not do it that way.,

**Comment by joachimmarder** [comment]
> I don't think we need to take a dependency on another project. 

Using a well maintained unit "Vcl.Styles.UxTheme" instead of our "VirtualTrees.StyleHooks" doesn't sound too bad to me, as long as it doesn't introduce new issues. It could be worth a try.

I never really understood why we need those more than 1100 lines of code (that were emailed to me years ago) to have styled scrollbars, although the other VCL controls already have them.

**Comment by pyscripter** [comment]
I really apologize.  The component shown in the pictures above is a TVirtualTree descendent (TVirtualExplorerTree from https://code.google.com/p/mustangpeakvirtualshelltools/).  

For the hooks to work you need to register the hooks for the class descendant.  Once this is done the scrollbars are themed properly without using Vcl.Styles.UxTheme.

I did get an exception (Invalid Windows handle) once from the VirtualTree style hook code (in my application I am using many VirtualTrees!), but I will open a new issue if this happens again.

So I am closing this issue.  And sorry for the trouble.

**Comment by pyscripter** [comment]
One suggestion might be to change the hook registration to:

``` Pascal
 TCustomStyleEngine.RegisterStyleHook(TCustomVirtualStringTree, TVclStyleScrollBarsHook);
 TCustomStyleEngine.RegisterStyleHook(TCustomVirtualDrawTree, TVclStyleScrollBarsHook);
```

so that descendant controls do not have to register the hooks.

**Comment by joachimmarder** [comment]
And let's say someone does not want this for a derived class? The current way is more flexible. And those people who derive from the VT classes to just add a few features typically derive typically derive from TVirtualStringTree and TVirtualStringTree.

**Comment by pyscripter** [comment]
By the way using the TMemoStyleHook from the Vcl Library standard unit Vcl.StdCtrls instead of the TVclStyleScrollBarsHook works just fine.  So it appears that there is no need to duplicate and maintain all this extra code.
i.e.
```delphi
 TCustomStyleEngine.RegisterStyleHook(TCustomVirtualStringTree, TMemoStyleHook );
 TCustomStyleEngine.RegisterStyleHook(TCustomVirtualDrawTree, TMemoStyleHook );
```
